tools/bisector/bisector/bisector.py

```diff
@@ -2558,6 +2563,105 @@ def upload(self, path):
 
         return path
 
+    def download(self, url, path):
+        """
+        Download a file from Artifactorial: anonymous access.
+
+        :param url: URL of the Artifactorial file to download
+        :param path: path to the downloaded file
+        """
+        urlretrieve(url, path)
+
+
+class ArtifactoryService(ArtifactsService):
+    """Upload/download files to/from Artifactory. """
+
+    def __init__(self, token=None):
+        """
+        :param token: Token granting the read & write access to the url,
+                      or env var ARTIFACTORIAL_TOKEN
+        """
+
+        self.token = token or os.getenv('ARTIFACTORIAL_TOKEN')
+
+        if not self.token:
+            raise ValueError("An Artifactory token must be provided through ARTIFACTORIAL_TOKEN env var")
+
+    def upload(self, path, url=None):
+        """
+        Upload a file to Artifactory.
+
+        :param path: path to the file to upload
+        :param url: URL of the Artifactory folder to upload to,
+                    or env var ARTIFACTORIAL_FOLDER
+        """
+
+        token = self.token
+
+        dest = url or os.getenv('ARTIFACTORIAL_FOLDER')
+        if not dest:
+            raise ValueError("An Artifactorial folder URL must be provided through ARTIFACTORIAL_FOLDER env var")
+
+        dest = dest + "/" + os.path.basename(path)
+
+        if not os.path.exists(path):
+            info(f'Cannot upload {path} as it does not exists.')
+            return path
+
+        info(f'Uploading {path} to {dest} ...')
+
+        with open(path, 'rb') as f:
+            content = f.read()
+            md5sum = hashlib.md5(content).hexdigest();
+
+        headers = {
+            "X-Checksum-Md5": md5sum,
+            "X-JFrog-Art-Api": token
+        }
+
+        response = requests.put(dest, headers=headers, data=content)
+
+        if response.ok:
+            url = response.json().get("downloadUri")
+            info('{path} uploaded to {url} ...'.format(
+                path=path,
+                url=url
+            ))
+            path = url
+        # Return the path unmodified if it failed
+        else:
+            warn('Failed to upload {path} to {dest} (HTTP {error}/{reason}): {msg}'.format(
+                path=path,
+                dest=dest,
+                error=response.status_code,
+                reason=response.reason,
+                msg=response.text,
+            ))
+
+        return path
+
+    def download(self, url, path):
+        """
+        Download a file from an Artifacts service.
+
+        :param path: path to the file to download
+        """
+
+        headers = {
+            "X-JFrog-Art-Api": self.token
+        }
+
+        response = requests.get(url, headers=headers)
+        # Raise an exception is the request failed
```

Comment:
Maybe we should just emit a warn() instead ? Raising an exception will prevent downloading any other link if one is dead for some reason. If that is the most convenient function from `requests`, we can simply:
```
try:
    ....
except Exception as e:
    warn(f'Fail to retrieve URL {url}: {e}')
```

Reply:
We do catch these exceptions when we call the download function, as in:

```
                          try:
                              download_service.download(url=archive_path,
                                                        path=archive_dst)
                          except Exception as e:
                              error('Could not retrieve {archive_path}: {e}'.format(
                                  archive_path=archive_path,
                                  e=e
                              ))
```

I'm tempted to either leave it as is, or modify the download functions for both the ArtifactorialService and the ArtifactoryService to catch and warn inside the download function. 

Any preferences?